pkg/logwrite/logwrite.go

```diff
@@ -0,0 +1,184 @@
+	Name         string   // filename of log file
+	Dir          string   // log file directory
+	BytesWritten int      // total number of bytes written so far
+}
+
+// NewLogFile creates a new LogFile
+func NewLogFile(dir, name string) (*LogFile, error) {
+	f, err := os.Create(filepath.Join(dir, name))
+	if err != nil {
+		return nil, err
+	}
+	return &LogFile{
+		File:         f,
+		Name:         name,
+		Dir:          dir,
+		BytesWritten: 0,
+	}, nil
+}
+
+// Write appends a message to the log file
+func (l *LogFile) Write(m *LogMessage) error {
+	s := m.String()
+	_, err := io.WriteString(l.File, s)
+	if err == nil {
+		l.BytesWritten += len(s)
+	}
+	return err
+}
+
+// Rotate closes the current log file, rotates the files and creates an
+// empty log file.
+func (l *LogFile) Rotate(maxLogFiles int) error {
+	if err := l.File.Close(); err != nil {
+		return err
+	}
+	path := filepath.Join(l.Dir, l.Name)
+	for i := maxLogFiles - 1; i >= 0; i-- {
+		newerFile := fmt.Sprintf("%s.%d", path, i-1)
+		// special case: if index is 0 we omit the suffix i.e. we expect
+		// foo foo.1 foo.2 up to foo.<maxLogFiles-1>
+		if i == 0 {
+			newerFile = path
+		}
+		olderFile := fmt.Sprintf("%s.%d", path, i)
+		// overwrite the olderFile with the newerFile
+		err := os.Rename(newerFile, olderFile)
+		if os.IsNotExist(err) {
+			// the newerFile does not exist
+			continue
+		}
+		if err != nil {
+			return err
+		}
+	}
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	l.File = f
+	l.BytesWritten = 0
+	return nil
+}
+
+func main() {
+	socketPath := flag.String("socket", "/var/run/memlogdq.sock", "memlogd log query socket")
+	logDir := flag.String("log-dir", "/var/log", "Directory containing log files")
+	maxLogFiles := flag.Int("max-log-files", 10, "Maximum number of rotated log files before deletion")
+	maxLogSize := flag.Int("max-log-size", 1048576, "Maximum size of a log file before rotation")
+	flag.Parse()
+
+	addr := net.UnixAddr{
+		Name: *socketPath,
+		Net:  "unix",
+	}
+	conn, err := net.DialUnix("unix", nil, &addr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	n, err := conn.Write([]byte{logDumpFollow})
+	if err != nil || n < 1 {
+		log.Fatalf("Failed to write request to memlogd socket: %s", err)
+	}
+
+	// map of service name to active log file
+	logs := make(map[string]*LogFile)
+
+	r := bufio.NewReader(conn)
+	for {
+		line, err := r.ReadString('\n')
+		if err == io.EOF {
+			return
+		}
+		if err != nil {
+			log.Fatalf("Failed to read from memlogd: %s", err)
+		}
+		msg, err := ParseLogMessage(line)
+		if err != nil {
+			log.Println(err)
+			continue
+		}
+		if strings.HasPrefix(msg.Name, "logwrite") {
+			// don't log our own output in a loop
+			continue
+		}
+
+		var logF *LogFile
+		var ok bool
+		if logF, ok = logs[msg.Name]; !ok {
+			logF, err = NewLogFile(*logDir, msg.Name)
+			if err != nil {
+				log.Fatalf("Failed to create log file %s: %s", msg.Name, err)
```

Comment:
This stops logging for all sources if *one* log file creation fails. Is this the failure mode we want?